File: test/CSharp/CodeCracker.Test/Refactoring/AllowMembersOrderingCodeFixProviderTests.Base.cs
using CodeCracker.CSharp.Refactoring;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using TestHelper;
using Xunit;

namespace CodeCracker.CSharp.Test.Refactoring
{
    public class BaseAllowMembersOrderingCodeFixProviderTests :
        CodeFixTest<AllowMembersOrderingAnalyzer, BaseAllowMembersOrderingCodeFixProviderTests.MockCodeFixProvider>
    {
        [Theory]
        [InlineData("class", "void B() { };", "void A() { };")]
        [InlineData("struct", "int c = 3, d = 4;", "int a = 1, b = 2;")]
        public async Task BaseAllowMembersOrderingShouldCallIComparerToOrder(string typeDeclaration, string memberA, string memberB)
        {
            var codeFixProvider = base.GetCSharpCodeFixProvider() as MockCodeFixProvider;

            var source = @"
            public " + typeDeclaration + @" Foo
            {
                " + memberA + @"
                " + memberB + @"
            }";

            var expected = @"
            public " + typeDeclaration + @" Foo
            {
                " + memberB + @"
                " + memberA + @"
            }";

            await VerifyCSharpFixAsync(source, expected, codeFixProvider: codeFixProvider);
            Assert.True(codeFixProvider.HasIComparerBeenCalled, "The IComparer must be used to sort the members of that type");
        }

        [Fact]
        public async Task BaseAllowMembersOrderingShouldNotRegisterFixIfIsAlreadySorted()
        {
            const string source = @"
            public class Foo
            {
                int a;
                void B() { }
            }";
            await VerifyCSharpHasNoFixAsync(source);
        }

        [Theory]
        [InlineData("class")]
        [InlineData("struct")]
        public async Task BaseAllowMembersOrderingShouldSupportWriteMembers(string typeDeclaration)
        {
            var source = @"
            using System;

            namespace ConsoleApplication1
            {
                " + typeDeclaration + @" Foo
                {
                    public class Foo2 { }
                    public struct Struct { }
                    public static Foo operator +(Foo f1, Foo f2) { return new Foo(); }
                    void Method(string a) { }
                    public string this[int i] { set { } }
                    public string Property { get; set; }
                    private interface Interface { }
                    public enum Enum { Enum1, Enum2 = 1 }
                    public event Action Event
                    {
                        add { EventField += value; }
                        remove { EventField -= value; }
                    }
                    public delegate double Delegate(double num);
                    public Foo()
                    {
                        Property = Field1 = Field = "";
                        EventField = EventField1 = () => { };
                    }
                    public event Action EventField1;
                    public string Field;
                }
            }";

            var expected = @"
            using System;

            namespace ConsoleApplication1
            {
                " + typeDeclaration + @" Foo
                {
                    private interface Interface { }
                    public class Foo2 { }
                    public delegate double Delegate(double num);
                    public enum Enum { Enum1, Enum2 = 1 }
                    public event Action Event
                    {
                        add { EventField += value; }
                        remove { EventField -= value; }
                    }
                    public event Action EventField1;
                    public Foo()
                    {
                        Property = Field1 = Field = "";
                        EventField = EventField1 = () => { };
                    }
                    public static Foo operator +(Foo f1, Foo f2) { return new Foo(); }
                    public string Field;
                    public string Property { get; set; }
                    public string this[int i] { set { } }
                    public struct Struct { }
                    void Method(string a) { }
                }
            }";

            await VerifyCSharpFixAsync(source, expected);
        }

        public class MockCodeFixProvider : BaseAllowMembersOrderingCodeFixProvider
        {
            public MockCodeFixProvider() : base("Fake codefix") { }

            public bool HasIComparerBeenCalled { get; private set; }

            protected override IComparer<MemberDeclarationSyntax> GetMemberDeclarationComparer(Document document, CancellationToken cancellationToken)
                => new AlphabeticalMemberOrderingComparer(this);

            internal class AlphabeticalMemberOrderingComparer : IComparer<MemberDeclarationSyntax>
            {
                readonly MockCodeFixProvider parent;

                public AlphabeticalMemberOrderingComparer(MockCodeFixProvider parent) { this.parent = parent; }

                public int Compare(MemberDeclarationSyntax x, MemberDeclarationSyntax y)
                {
                    parent.HasIComparerBeenCalled = true;
                    return string.Compare(x.ToFullString(), y.ToFullString(), System.StringComparison.InvariantCulture);
                }
            }
        }

    }
}

File: test/VisualBasic/CodeCracker.Test/Refactoring/AllowMembersOrderingCodeFixProviderTests.Base.vb
Imports System.Threading
Imports CodeCracker.VisualBasic.Refactoring
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Xunit

Namespace Refactoring
    Public Class BaseAllowMembersOrderingCodeFixProviderTests
        Inherits CodeFixVerifier(Of AllowMembersOrderingAnalyzer, MockCodeFixProvider)

        <Theory>
        <InlineData("Class", "Property B As String", "Property A As String")>
        Public Async Function BaseAllowMembersOrderingShouldCallIComparerToOrder(typeDeclaration As String, memberA As String, memberB As String) As Task

            Dim codeFixProvider = DirectCast(MyBase.GetCodeFixProvider, MockCodeFixProvider)


            Dim source = String.Format("
Public {0} Foo
    {1}
    {2}
End {0}", typeDeclaration, memberA, memberB)

            Dim expected = String.Format("
Public {0} Foo
    {2}
    {1}
End {0}", typeDeclaration, memberA, memberB)

            Await VerifyBasicFixAsync(source, expected, codeFixProvider:=codeFixProvider)
            Assert.True(codeFixProvider.HasIComparerBeenCalled, "The IComparer must be used to sort the members of that type")
        End Function

        <Theory>
        <InlineData("Class")>
        <InlineData("Structure")>
        Public Async Function BaseAllowMembersOrderingShouldSupportWriteMembers(typeDeclaration As String) As Task
            Dim source = String.Format("
Imports System

Namespace ConsoleApplication1
    {0} Foo
        Public Class Foo2
        End Class
        Public Structure Struct

        End Structure
        Public Shared Operator +(f1 As Foo, f2 As Foo) As Foo
            Return New Foo()
        End Operator
        Sub Method(a As String)

        End Sub
        Public Property Prop As String
        Private Interface IInterface

        End Interface
        Public Enum Eenum
            Enum1
            Enum2 = 1
        End Enum
        Public Custom Event CustomEvent As EventHandler
            AddHandler(value As EventHandler)

            End AddHandler
            RemoveHandler(value As EventHandler)

            End RemoveHandler
            RaiseEvent(sender As Object, e As EventArgs)

            End RaiseEvent
        End Event
        Public Delegate Function doubleDelegate(num As Double) As Double
        Public Sub New()
            Prop = """"
            Field = """"
        End Sub
        Public Event EventField1 As Action
        Public field As String
    End {0}
End Namespace
", typeDeclaration)

            Dim expected = String.Format("
Imports System

Namespace ConsoleApplication1
    {0} Foo
        Private Interface IInterface

        End Interface
        Public Class Foo2
        End Class
        Public Custom Event CustomEvent As EventHandler
            AddHandler(value As EventHandler)

            End AddHandler
            RemoveHandler(value As EventHandler)

            End RemoveHandler
            RaiseEvent(sender As Object, e As EventArgs)

            End RaiseEvent
        End Event
        Public Delegate Function doubleDelegate(num As Double) As Double
        Public Enum Eenum
            Enum1
            Enum2 = 1
        End Enum
        Public Event EventField1 As Action
        Public field As String
        Public Property Prop As String
        Public Shared Operator +(f1 As Foo, f2 As Foo) As Foo
            Return New Foo()
        End Operator
        Public Structure Struct

        End Structure
        Public Sub New()
            Prop = """"
            Field = """"
        End Sub
        Sub Method(a As String)

        End Sub
    End {0}
End Namespace
", typeDeclaration)

            Await VerifyBasicFixAsync(source, expected)
        End Function
        Public Class MockCodeFixProvider
            Inherits BaseAllowMembersOrderingCodeFixProvider
            Public Sub New()
                MyBase.New("Fake codefix")
            End Sub

            Public Property HasIComparerBeenCalled As Boolean

            Protected Overrides Function GetMemberDeclarationComparer(document As Document, cancellationToken As CancellationToken) As IComparer(Of DeclarationStatementSyntax)
                Return New AlphabeticalMemberOrderingComparer(Me)
            End Function
            Friend Class AlphabeticalMemberOrderingComparer
                Implements IComparer(Of DeclarationStatementSyntax)

                ReadOnly parent As MockCodeFixProvider

                Public Sub New(parent As MockCodeFixProvider)
                    Me.parent = parent
                End Sub

                Public Function Compare(x As DeclarationStatementSyntax, y As DeclarationStatementSyntax) As Integer Implements IComparer(Of DeclarationStatementSyntax).Compare
                    parent.HasIComparerBeenCalled = True
                    Return String.Compare(x.ToFullString, y.ToFullString, StringComparison.InvariantCulture)
                End Function
            End Class
        End Class
    End Class
End Namespace